master/buildbot/reporters/slack.py

```diff
@@ -0,0 +1,177 @@
+                isinstance(default_channels, (list, tuple))):
+            config.error('default_channels parameter must be a list or a tuple')
+
+        super(SlackReporter).checkConfig(*args, **kwargs)
+
+    @defer.inlineCallbacks
+    def reconfigService(self, username, url, mapping, default_channels,
+                        *args, **kwargs):
+        """
+        reconfigure the service based on the new parameters
+        """
+        yield super(SlackReporter).reconfigService(self, *args, **kwargs)
+
+        # pylint: disable=attribute-defined-outside-init
+
+        self._username = username
+
+        # TODO: this goes to the documentation
+        # mapping is expected to have the following structure
+        # mapping = {
+        #     'builder1': {
+        #         'kinds': ('start',),
+        #         'channels': ('#channel1', '#channel2')
+        #     },
+        #     'builder2': {
+        #         'kinds': ('finish',),
+        #     },
+        #     'builder3': {
+        #         'kinds': ('start', 'finish')
+        #     },
+        # }
+        self._mapping = mapping
+        # TODO: this goes to the documentation
+        # list of channels to post a message to if none are specified in the
+        # mapping.  the default behaviour is to use the channel configured for
+        # the webhook
+        self._default_channels = default_channels
+
+        self._http = yield HTTPClientService.getService(self.master, url)
+
+    def _filter_builds(self, build):
+        """
+        check if we should post a message for the given build:
+
+        * the builder is present in the map
+        * 'kinds' has the value for the kind of this notification (start/finish)
+        """
+        not_interested = None, None
+
+        props = self._mapping.get(build['builder'], dict())
+        if not props:
+            return not_interested
+
+        kind = 'finish' if build['complete'] else 'start'
+        if kind not in props['kinds']:
+            return not_interested
+
+        if self._default_channels is not _Default:
+            channels = props.get('channels', self._default_channels)
+            if not channels:
+                return not_interested
+        else:
+            channels = None
+
+        return props, channels
+
+    @staticmethod
+    def _prepare_attachment(build):
+        """
+        prepare an attachment based on the build information
+        """
+        if build['complete']:
+            result = build['results']
+
+            title = 'Build #{} has finished: {}'.format(build['buildid'],
+                                                        result)
+            # NOTE: we might want to include links to the logs of failed
+            # step(s) to 'text' => update neededDetails for the class
+            text = ' '.join(build.getText())
+            timestamp = build['complete_at']
+            color = 'good' if result == SUCCESS else 'danger'
```

Comment:
is SKIP or WARNING results still a danger?

usually we implement a full dictionnary to map all possible values for results